### PR TITLE
Filter out those facilities that are already linked to user

### LINKED ISSUE #1581: Don't show already Linked Facilities on Link New Facility on User page
**Describe the bug**
When a user tries to link a facility to the user, the already linked facilities are also shown.

**To Reproduce**

1. Go to https://care.coronasafe.in/users
2. Click on `Linked Facility`
3. In the form, add a facility that is already linked and click `Add`

**Expected behavior**
Only those facilities that are not linked to the user should be displayed.

**Screenshots**
image
image

**Desktop (please complete the following information):**

**Smartphone (please complete the following information):**

**Additional context**

### PR DESCRIPTION
**Pull request by Pranshu1902**
## Proposed Changes

- Filter out those facilities that are already linked to the user. When the Link Facility page makes a call to the backend, if the username is passed then the backend will filter out those facilities that are already linked.

### Associated Issue
Fixes #1581 

## Merge Checklist

- [ ] Tests added/fixed
- [ ] Update docs in `/docs`
- [ ] Linting Complete
- [ ] Any other necessary step

_*Only PR's with test cases included and passing lint and test pipelines will be reviewed*_

@coronasafe/care-backend-maintainers @coronasafe/care-backend-admins

### CODE REVIEW

**Comment by nihal467** [comment]
@Pranshu1902 clear the failed test

File: care/facility/api/viewsets/facility.py [hunk 1]
```diff
@@ -39,6 +41,14 @@ class FacilityFilter(filters.FilterSet):
     state = filters.NumberFilter(field_name="state__id")
     state_name = filters.CharFilter(field_name="state__name", lookup_expr="icontains")
     kasp_empanelled = filters.BooleanFilter(field_name="kasp_empanelled")
+    exclude_user = filters.BooleanFilter(method='filter_exclude_user')
+
+    def filter_exclude_user(self, queryset, name, value):
+        if value:
+            user = self.request.query_params.get('exclude_user')
+            if user:
+                queryset = queryset.exclude(facilityuser__user__username=user)
```
Comment (sainak) [inline care/facility/api/viewsets/facility.py:50]:
```suggestion
            queryset = queryset.exclude(facilityuser__user__username=value)
```
value is the vlaue you are looking for

File: care/facility/api/viewsets/facility.py [hunk 2]
```diff
@@ -95,6 +105,15 @@ def get_parsers(self):
...
 
+    def get_queryset(self):
+        queryset = super().get_queryset()
+        exclude_user_param = self.request.query_params.get("exclude_user")
+
+        if exclude_user_param:
+            queryset = queryset.exclude(facilityuser__user__username=exclude_user_param)
+
+        return queryset
+
```
Comment (sainak) [inline care/facility/api/viewsets/facility.py:116]:
```suggestion
```
this is not required, as the filter will be applied from the FacilityFilterset

File: care/facility/tests/test_facilityuser_api.py [hunk 3]
```diff
@@ -25,3 +25,28 @@ def test_get_queryset_with_prefetching(self):
...
+
+    def test_link_existing_facility(self):
+        print(self.users[0])
+        response = self.new_request(
+            (f"/api/v1/getallfacilities",),
+            {"get": "list", "exclude_user": self.users[0].username},
+            FacilityViewSet,
+            self.users[0],
+            {"facility_external_id": self.facility.external_id},
+        )
```
Comment (sainak) [inline care/facility/tests/test_facilityuser_api.py:49]: use self.client.get instead of self.new_request

**Review by sainak — CHANGES_REQUESTED** [review]

File: care/facility/tests/test_facilityuser_api.py (hunk 3 above)
Comment (Pranshu1902) [reply care/facility/tests/test_facilityuser_api.py]: Updated

**Comment by nihal467** [comment]
@Pranshu1902 the test is still failing

**Comment by nihal467** [comment]
@sainak review the backend, the functionality looks good to me

**Review by rithviknishad — COMMENTED** [review]
@Pranshu1902 are you sure pre-commit lint is working fine in your local?

Check the linting CI

**Review by sainak — APPROVED** [review]

**Review by rithviknishad — APPROVED** [review]

**Comment by nihal467** [comment]
tested locally, looks good to me

**Pull request merged**